clarity/contracts/arkadiko-liquidation-rewards-v1-2.clar

```diff
@@ -0,0 +1,232 @@
+(define-constant ERR-TOKEN-NOT-WHITELISTED u30004)
+(define-constant ERR-REWARD-LOCKED u30005)
+
+;; Variables
+(define-data-var total-reward-ids uint u0)
+(define-data-var shutdown-activated bool false)
+
+;; ---------------------------------------------------------
+;; Maps
+;; ---------------------------------------------------------
+
+(define-map reward-data
+  { 
+    reward-id: uint,
+  } 
+  {
+    share-block: uint, ;; STX block on which USDA shares are calculated
+    unlock-block: uint, ;; BTC block on which stacking tokens unlock
+    token-is-stx: bool,
+    token: principal,
+    total-amount: uint
+  }
+)
+
+(define-map reward-claimed
+  { 
+    reward-id: uint,
+    user: principal
+  } 
+  {
+    claimed: bool
+  }
+)
+
+(define-map reward-tokens
+  { 
+    token: principal,
+  } 
+  {
+    whitelisted: bool
+  }
+)
+
+(define-read-only (get-total-reward-ids)
+  (var-get total-reward-ids)
+)
+
+;; @desc get data for given reward
+;; @param reward-id; the reward
+(define-read-only (get-reward-data (reward-id uint))
+  (default-to
+    { share-block: u0, unlock-block: u0, token-is-stx: false, token: .usda-token, total-amount: u0 }
+    (map-get? reward-data { reward-id : reward-id  })
+  )
+)
+
+;; @desc get amount a user has already claimed for given reward
+;; @param reward-id; the reward
+;; @param user; the user who claimed
+(define-read-only (get-reward-claimed (reward-id uint) (user principal))
+  (default-to
+    { claimed: false }
+    (map-get? reward-claimed { reward-id : reward-id, user: user })
+  )
+)
+
+;; @desc get reward data combined with user info
+;; @param reward-id; the reward
+(define-public (get-user-reward-info (reward-id uint) (user principal) (liquidation-pool <liquidation-pool-trait>))
+  (let (
+    (sender-rewards (unwrap-panic (get-rewards-of user reward-id liquidation-pool)))
+    (rewards-data (get-reward-data reward-id))
+  )
+    (ok {
+      reward-id: reward-id,
+      pending-rewards: sender-rewards,
+      unlock-block: (get unlock-block rewards-data),
+      token: (get token rewards-data),
+      token-is-stx: (get token-is-stx rewards-data)
+    })
+  )
+)
+
+(define-read-only (token-whitelisted (token principal))
+  (get whitelisted (default-to
+    { whitelisted: false }
+    (map-get? reward-tokens { token: token  })
+  ))
+)
+
+(define-read-only (get-shutdown-activated)
+  (or
+    (unwrap-panic (contract-call? .arkadiko-dao get-emergency-shutdown-activated))
+    (var-get shutdown-activated)
+  )
+)
+
+;; ---------------------------------------------------------
+;; Add rewards
+;; ---------------------------------------------------------
+
+(define-public (add-reward (share-block uint) (token-is-stx bool) (token <ft-trait>) (total-amount uint))
+  (add-reward-locked share-block u0 token-is-stx token total-amount)
+)
+
+(define-public (add-reward-locked (share-block uint) (unlock-block uint) (token-is-stx bool) (token <ft-trait>) (total-amount uint))
```

Comment:
Q: this is usually only called from the auction engine, right?

Reply:
A: correct